File: backend/app/api/endpoints/voice.py
"""
Voice API endpoints.
"""
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Form, Response
from typing import Optional
from app.api.dependencies.auth import require_authenticated_user
from app.services.voice.voice_service import voice_service
from app.models.voice import ASRRequest, TTSRequest
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/voice", tags=["voice"])

@router.post("/asr")
async def transcribe_audio(
    interview_id: str = Form(...),
    language: str = Form(...),
    file: UploadFile = File(...),
    current_user: dict = Depends(require_authenticated_user),
):
    """
    Transcribe audio using ASR provider.
    """
    # Read audio data
    audio_data = await file.read()
    mime_type = file.content_type or "audio/wav"
    
    result = await voice_service.transcribe_audio(
        current_user,
        interview_id,
        audio_data,
        language,
        mime_type,
    )
    
    if not result["success"]:
        raise HTTPException(
            status_code=result.get("status_code", 400),
            detail=result["error"],
        )
    
    return {
        "success": True,
        "data": result["transcription"],
    }

@router.post("/tts")
async def synthesize_speech(
    request: TTSRequest,
    current_user: dict = Depends(require_authenticated_user),
):
    """
    Synthesize text to speech.
    Returns audio data.
    """
    result = await voice_service.synthesize_speech(
        current_user,
        request.interview_id,
        request.text,
        request.language,
        request.voice_id,
    )
    
    if not result["success"]:
        raise HTTPException(
            status_code=result.get("status_code", 400),
            detail=result["error"],
        )
    
    audio = result["audio"]
    return Response(
        content=audio["data"],
        media_type=audio["mime_type"],
        headers={"X-Provider": audio["provider"]},
    )

File: frontend/src/services/voice.js
const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || '/api';

class VoiceService {
  getToken() {
    return sessionStorage.getItem('medikiosk_access_token');
  }
  
  async transcribeAudio(interviewId, language, audioBlob, mimeType = 'audio/wav') {
    const token = this.getToken();
    if (!token) return { success: false, error: 'Not authenticated' };
    
    try {
      const formData = new FormData();
      formData.append('interview_id', interviewId);
      formData.append('language', language);
      formData.append('file', audioBlob, `audio.${mimeType.split('/')[1] || 'wav'}`);
      
      const response = await fetch(`${API_BASE_URL}/voice/asr`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${token}`,
        },
        body: formData,
      });
      
      const data = await response.json();
      
      if (!response.ok) {
        return { success: false, error: data.detail || 'Transcription failed' };
      }
      
      return { success: true, transcription: data.data };
    } catch (error) {
      return { success: false, error: 'Network error. Voice transcription unavailable.' };
    }
  }
  
  async synthesizeSpeech(interviewId, text, language, voiceId = null) {
    const token = this.getToken();
    if (!token) return { success: false, error: 'Not authenticated' };
    
    try {
      const response = await fetch(`${API_BASE_URL}/voice/tts`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          interview_id: interviewId,
          text,
          language,
          voice_id: voiceId,
        }),
      });
      
      if (!response.ok) {
        const data = await response.json();
        return { success: false, error: data.detail || 'Speech synthesis failed' };
      }
      
      const audioBlob = await response.blob();
      const audioUrl = URL.createObjectURL(audioBlob);
      
      return { success: true, audioUrl };
    } catch (error) {
      return { success: false, error: 'Network error. Voice playback unavailable.' };
    }
  }
}

export const voiceService = new VoiceService();